src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. 
+
+using System;
+using System.Linq;
+using System.Collections;
+using System.Collections.Generic;
+using Microsoft.OpenApi.Exceptions;
+using Microsoft.OpenApi.Properties;
+
+namespace Microsoft.OpenApi.Validations.Rules
+{
+    /// <summary>
+    /// The rule set of the validation.
+    /// </summary>
+    public sealed class ValidationRuleSet : IEnumerable<ValidationRule>
+    {
+        private IDictionary<Type, IList<ValidationRule>> _rules = new Dictionary<Type, IList<ValidationRule>>();
+
+        /// <summary>
+        /// The default rule set.
+        /// </summary>
+        public static ValidationRuleSet DefaultRuleSet = new ValidationRuleSet
+        {
+            OpenApiComponentsRules.KeyMustBeRegularExpression,
+
+            OpenApiDocumentRules.FieldIsRequired,
+
+            OpenApiInfoRules.FieldIsRequired,
+
+            OpenApiLicenseRules.FieldIsRequired,
+
+            OpenApiContactRules.EmailMustBeEmailFormat,
+
+            OpenApiExternalDocsRules.FieldIsRequired,
+
+            OpenApiServerRules.FieldIsRequired,
+
+            OpenApiTagRules.FieldIsRequired,
+
+            OpenApiResponseRules.FieldIsRequired,
+
+            OpenApiOAuthFlowRules.FieldIsRequired,
+
+            OpenApiExtensibleRules.ExtensionNameMustStartWithXDash,
+            // add more default rules.
+        };
```

Comment:
This will be difficult to track. People will forget to put the new rules in here since this list is in a different file than the rules themselves.

I am thinking of this:

- Add DefaultRuleSet for every OpenApiXXXRules file
- This main DefaultRuleSet is a concatenation of all DefaultRuleSet's

What do you think?

Reply:
In fact, we might not need this concatenated list at all since it will be broken down in Add() below anyway.

---
In reply to: [157320269](https://github.com/Microsoft/OpenAPI.NET/pull/152#discussion_r157320269) [](ancestors = 157320269)

Reply:
1. Basically, I want to use the attribute for each "DOM" classes same as #133. 
    Let me think it more.

2. Why do you think it will break the Add() method? That's in deed will call "Add()" method.

Reply:
If you wanted to create an OpenApiValidationRuleAttribute and annotate each rule with that attribute and then have a discovery mechanism that searched the assembly to find all classes with that attribute, then I would have no objection to that.

---
In reply to: [157321750](https://github.com/Microsoft/OpenAPI.NET/pull/152#discussion_r157321750) [](ancestors = 157321750)

Reply:
Overall I like the direction of this PR.  I think by adding a simple marker attribute to each rule, you can reflect over the rules and built a default ruleset.  You should be able to determine the target type from the generic return type of the rule.

This should address the concern brought up by @PerthCharern  that someone might forget to add the rule to the ruleset.

Reply:
@xuzhg I didn't mean it will break the Add() method. I meant the list didn't have to be created because it would be enumerated in the Add() method anyway, so we can find a way to make it less complicated.

Anyway, I like @darrelmiller 's suggestion here. That would put the concern of when each rule should be used back to the rule itself.

Reply:
Changed the design to use reflect. 

---
In reply to: [157320436](https://github.com/Microsoft/OpenAPI.NET/pull/152#discussion_r157320436) [](ancestors = 157320436,157320269)

Reply:
Changed.

---
In reply to: [157326026](https://github.com/Microsoft/OpenAPI.NET/pull/152#discussion_r157326026) [](ancestors = 157326026,157321750)

Reply:
Changed.

---
In reply to: [157327739](https://github.com/Microsoft/OpenAPI.NET/pull/152#discussion_r157327739) [](ancestors = 157327739)